### PR TITLE
bank: deposit and withdraw functionality

### LINKED ISSUE #71: bank: multiple users should be able to deposit and withdraw
Multiple users should be able to deposit and withdraw money from accounts.

This could be made by supporting log-in functionality for authentication and separation of bank accounts and there transactions on a user level.

### LINKED ISSUE #75: bank: deposit and withdraw from personal account
We should implement deposit and withdraw functionality.
Given that the user is logged in he should be able to withdraw and deposit from the amount in his account.

Tasks
- Withdraw and deposit functionality on the html page.
- Saving the data in to the database.

### PR DESCRIPTION
**Pull request by gadjevb**
Added deposit and withdraw functionality, so that users can operate with their money . Username is hardcoded because there's still no session to get it from.

Fixes #75 
Fixes #71

account

### CODE REVIEW

File: src/main/java/com/clouway/http/servlets/HomeServlet.java [hunk 1]
```diff
@@ -1,21 +0,0 @@
-package com.clouway.http.servlets;
```
Comment (mgenov) [inline src/main/java/com/clouway/http/servlets/HomeServlet.java]: This is the page where user goes after login. I suppose "/account" in this PR should be moved in here.

File: src/main/java/com/clouway/http/servlets/TransactionPageServlet.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    Optional<Account> account = repository.getByName("John");// todo get username from session
+
+    if (!account.isPresent()) {
+      // todo sendRedirect to main page
+    }
+
+    values.put("name", account.get().name);
+    values.put("error", "");
+
+    if (type.equals("deposit")) {
```
Comment (mgenov) [inline src/main/java/com/clouway/http/servlets/TransactionPageServlet.java]:
```java
if ("deposit".equals(type))  {
..
} else if ("withdraw".equals(type)) {
..
} else {
   throw new IllegalStateException(String.format("operationType = '%s' is not currently supported", type)
}
```

**Comment by freekman** [comment]
@gadjevb  again missing UI

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java [hunk 3]
```diff
@@ -36,6 +36,18 @@ public void happyPath() throws Exception {
   }
 
   @Test
+  public void update() throws Exception {
```
Comment (freekman) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: by looking at the name of the test can you guess what is the test doing

File: src/main/webapp/WEB-INF/web.xml [hunk 4]
```diff
@@ -47,6 +52,11 @@
   </servlet-mapping>
 
   <servlet-mapping>
+    <servlet-name>transactionPageServlet</servlet-name>
```
Comment (freekman) [inline src/main/webapp/WEB-INF/web.xml]: the url-pattern is ```/account``` ,so the name of the servlet could also be AccountServlet or something like that

File: src/main/java/com/clouway/http/servlets/HomeServlet.java (hunk 1 above)
Comment (gadjevb) [inline src/main/java/com/clouway/http/servlets/HomeServlet.java]: The page after login is HomePageServlet

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java [hunk 5]
```diff
@@ -36,6 +36,18 @@ public void happyPath() throws Exception {
...
+    Account account = new Account("John", "123", 123);
+    Account updated = new Account("John", "123", 500);
+
+    repo.register(account);
+    repo.updateBalance(updated.name, updated.amount);
```
Comment (gadjevb) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: Fixed!

File: src/main/webapp/WEB-INF/web.xml (hunk 4 above)
Comment (gadjevb) [inline src/main/webapp/WEB-INF/web.xml]: Fixed!

File: src/main/java/com/clouway/http/servlets/TransactionPageServlet.java (hunk 2 above)
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/TransactionPageServlet.java]: @gadjevb  this is still no done

File: src/main/java/com/clouway/http/servlets/AccountPageServlet.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    String type = req.getParameter("transaction");
+    Integer amount = Integer.parseInt(req.getParameter("amount"));
+
+    Optional<Account> account = repository.getByName("John");// todo get username from session
+
+    if (!account.isPresent()) {
+      // todo sendRedirect to main page
+    }
+
+    values.put("name", account.get().name);
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/AccountPageServlet.java]: account.get().... is duplicated several times

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepository.java [hunk 7]
```diff
@@ -28,13 +31,39 @@ public void register(Account account) {
...
   @Override
+  public void deposit(String name, Double balance) {
+    dataStore.runInTransaction(new TransactionCall() {
+      @Override
+      public void execute(Connection connection) {
+        BigDecimal currentBalance = dataStore.fetchOne(connection, "select amount from accounts where name=?", name);
+        dataStore.executeUpdate(connection, "update accounts set amount=? where name=?", currentBalance.add(BigDecimal.valueOf(balance)), name);
+      }
+    });
+    //String query = "update accounts set amount=amount+? where name=?;";
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepository.java]: .

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepository.java [hunk 8]
```diff
@@ -28,13 +31,39 @@ public void register(Account account) {
...
+  @Override
+  public void withdraw(String name, Double balance) {
+    dataStore.runInTransaction(new TransactionCall() {
+      @Override
+      public void execute(Connection connection) {
+        BigDecimal currentBalance = dataStore.fetchOne(connection, "select amount from accounts where name=?", name);
+        dataStore.executeUpdate(connection, "update accounts set amount=? where name=?", currentBalance.subtract(BigDecimal.valueOf(balance)), name);
+      }
+    });
+    //String query = "update accounts set amount=amount+? where name=?;";
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentAccountRepository.java]: .

**Comment by freekman** [comment]
LGTM after fix comments i would like for you to discuss who will merge first you or @martinmilev

**Comment by gadjevb** [comment]
@freekman comments are fixed, I will wait for Martin to merge first. It will be easier to fix the conflicts from my branch.

**Pull request merged**